web3.js/src/connection.ts

```diff
@@ -3154,6 +3176,143 @@ export class Connection {
     return result;
   }
 
+  private async confirmTransactionUsingDurableNonceStrategy({
+    commitment,
+    signature,
+    strategy: {minContextSlot, nonceAccountPubkey, nonceValue},
+  }: {
+    commitment?: Commitment;
+    signature: string;
+    strategy: NonceBasedTransactionConfirmationStrategy;
+  }) {
+    let done: boolean = false;
+    const expiryPromise = new Promise<{
+      __type: TransactionStatus.NONCE_INVALID;
+    }>(resolve => {
+      const INVALID_NONCE_ERROR_SENTINEL = {} as const;
+      let currentNonceValue = nonceValue;
+      const getCurrentNonceValue = async () => {
+        try {
+          const nonceAccount = await this.getNonce(nonceAccountPubkey, {
+            commitment,
+            minContextSlot,
+          });
+          if (nonceAccount == null) {
+            throw INVALID_NONCE_ERROR_SENTINEL;
+          }
+          return nonceAccount.nonce;
+        } catch (e) {
+          if (e === INVALID_NONCE_ERROR_SENTINEL) {
+            throw e;
+          }
+          // If for whatever reason we can't reach/read the nonce
+          // account, just keep using the last-known value.
+          return currentNonceValue;
+        }
+      };
+      (async () => {
+        try {
+          currentNonceValue = await getCurrentNonceValue();
+          if (done) return;
+          while (
+            true // eslint-disable-line no-constant-condition
+          ) {
+            if (nonceValue !== currentNonceValue) {
+              resolve({__type: TransactionStatus.NONCE_INVALID});
+              return;
+            }
+            await sleep(2000);
+            if (done) return;
+            currentNonceValue = await getCurrentNonceValue();
+            if (done) return;
+          }
+        } catch (e) {
+          if (e === INVALID_NONCE_ERROR_SENTINEL) {
+            resolve({__type: TransactionStatus.NONCE_INVALID});
+            return;
+          } else {
+            throw e;
+          }
+        }
+      })();
+    });
+    const [confirmationPromise, cancelConfirmationDetector] =
+      this.getTransactionConfirmationPromise({commitment, signature});
+    let result: RpcResponseAndContext<SignatureResult>;
+    try {
+      const outcome = await Promise.race([confirmationPromise, expiryPromise]);
+      if (outcome.__type === TransactionStatus.PROCESSED) {
+        result = outcome.response;
+      } else {
+        // Double check that the transaction is indeed unconfirmed.
```

Comment:
The following code should probably get its own function

Reply:
I gave this a shot, and all it resulted in was some awkward input/output typing, strange function names, and the developer having to hop between functions to figure out what's going on. I think that this functionality is specific enough to durable nonce transactions such that breaking it into a separate function just adds friction without adding any of the good stuff, like reusability.

Reply:
I was imagining this being a function that ensures that a transaction wasn't processed before a particular context slot. I think that could be broadly useful (the blockheight exceedance strategy should use it as well).